### DISCUSSION TITLE
Deeply nested subresources not supported?

### DISCUSSION THREAD

**Discussion by usu** [api-platform/core#5673] [Q&A] [answered]
I was following the documentation on [Subresources](https://api-platform.com/docs/core/subresources/). All the examples given in the documentation have 1 level of nesting only. It seems, however that multiple Links are also supported. The following resource configuration works for me:

```php
#[ApiResource(
    uriTemplate: '/foo/{fooId}/bar/{barId}/baz.{_format}',
    uriVariables: [
        'fooId' => new Link(
            fromClass: Foo::class,
            toClass: Bar::class,
            toProperty: 'foo'
        ),
        'barId' => new Link(
            fromClass: Bar::class,
            toProperty: 'bar'
        ),
    ],
    operations: [
        new GetCollection(),
    ]
)]
class Baz { ... }
```

This seems to work. When requesting `GET /foo/1/bar/2/baz` I only receive `Baz` entities that belong to `/bar/2` and only so, if `/bar/2` belongs to `/foo/1`.

What I was wondering is, whether I can avoid the bar uriVariable (requesting `GET /foo/1/baz`). I was trying the configuration below, but didn't succeed. Not sure, if this is even supported at the moment:

```php
#[ApiResource(
    uriTemplate: '/foo/{fooId}/baz.{_format}',
    uriVariables: [
        'fooId' => new Link(
            fromClass: Foo::class,
            toClass: Bar::class,
            toProperty: 'foo'
        ),
        'barId' => new Link(
            fromClass: Bar::class,
            toProperty: 'bar'
        ),
    ],
    operations: [
        new GetCollection(),
    ]
)]
class Baz { ... }
```

Above configuration returns a ["Invalid identifier value or configuration" error in ReadListener](https://github.com/api-platform/core/blob/27f20969f532ebd748c11ad15d8e57366325f4cf/src/Symfony/EventListener/ReadListener.php#L99).

As an alternative, I'm also happy to provide my own DQL directly from the Doctrine Repositories. However, I couldn't find a way to specify a uriVariable that is used for routing only but ignored otherwise from Api-Platform. Either ReadListener is not happy, because I use a uriVariable that is not known to it. Or LinksHandlerTrait compiles an invalid DQL.

**Comment by usu** [comment, marked as answer]
Looking at the code of `LinksHandlerTrait`, I figured I can use `expandedValue` to have `LinksHandlerTrait` ignoring the uri variable. Below code seems to work - I don't know why, though 🤣 

`fooId` is properly populated into `$context['uri_variables']` and I can use a `QueryCollectionExtensionInterface` to filter the Baz collection by `fooId`:

```php
#[ApiResource(
    uriTemplate: '/foo/{fooId}/baz.{_format}',
    uriVariables: [
        'fooId' => new Link(
            fromClass: Foo::class,
            expandedValue: "{fooId}"
        ),
    ],
    operations: [
        new GetCollection(),
    ]
)]
class Baz { ... }
```

The only issue is, that fooId will not appear in OpenAPI documentation as available parameter.

I couldn't find any documentation on `expandedValue`, though. What is it's intention to do?

Is above code supposed to work? Or is this just a glitch?

**Reply by TNAJanssen** [reply]
Yeah same conclusion here, but i think the links logic should be rethought about as the working right now is weird by using the previous alias.

**Reply by soyuka** [reply]
Exactly, this is why we introduced `expandedValue` it just skips some of the logic. This is quite hard because URI variables: 
- handle how to get your identifiers from the objects (Object => IRI)
- also handle how to retrieve the data (IRI => Object)

For the first one if we miss the middle part of the chain we can't "guess" and you need to handle that yourself. If you find a way to explain this and think it's worth documenting I'd be glad to accept a PR on the documentation. I'm not a huge fan of the name, nor how it's currently coded I think there's room for improvement. 

For reference it is "documented" at : https://github.com/api-platform/core/blob/e895b69a04cb52eb6eb030bcc51a1e143dd357cb/docs/adr/0004-link.md#decision-outcome
